File: backend/src/Modules/Identity/Identity.Application/Users/Login/LoginHandler.cs
using FluentValidation;
using Identity.Application.Auth;
using Identity.Application.Contracts;
using Identity.Application.Validation;
using Identity.Domain;
using Identity.Domain.Tenants;
using SharedKernel;

namespace Identity.Application.Users.Login;

/// <summary>Kullanıcı giriş işlemini yürüten handler.</summary>
public sealed class LoginHandler(
    IValidator<LoginCommand> validator,
    IUserRepository users,
    ITenantMembershipRepository memberships,
    ITenantRepository tenants,
    IPasswordService passwords,
    ITokenService tokens) : ILoginHandler
{
    /// <inheritdoc/>
    public async Task<Result<LoginResult>> ExecuteAsync(
        LoginCommand command,
        CancellationToken cancellationToken = default)
    {
        var validationResult = await validator.ValidateToResultAsync(command, cancellationToken);
        if (validationResult.IsFailure)
        {
            return Result.Failure<LoginResult>(validationResult.Error);
        }

        var user = await users.GetByEmailAsync(command.Email.Trim().ToLowerInvariant(), cancellationToken);
        if (user is null || !passwords.VerifyPassword(user, command.Password, user.PasswordHash))
        {
            return Result.Failure<LoginResult>(Error.Unauthorized("Invalid credentials."));
        }

        var membership = await memberships.GetPrimaryForUserAsync(user.Id, cancellationToken);
        if (membership is null)
        {
            return Result.Failure<LoginResult>(Error.Unauthorized("User is not assigned to a tenant."));
        }

        var tenant = await tenants.GetByIdAsync(membership.TenantId, cancellationToken);
        if (tenant is null)
        {
            return Result.Failure<LoginResult>(Error.Unauthorized("Tenant not found."));
        }

        var (token, expiresAt) = tokens.GenerateToken(user, tenant);
        return Result.Success(new LoginResult(
            token,
            expiresAt,
            new UserDto(user.Id, user.Email, user.Name),
            new TenantDto(tenant.Id, tenant.Name)));
    }
}

File: backend-go/internal/modules/identity/application/login_handler.go
package application

import (
	"context"
	"strings"

	"pimly.commerslab/backend-go/internal/modules/identity/domain/users"
	"pimly.commerslab/backend-go/internal/sharedkernel"
)

// LoginCommand, kullanıcı giriş isteğini taşır.
type LoginCommand struct {
	Email    string
	Password string
}

// LoginHandler, kullanıcı giriş işlemini yürütür (.NET LoginHandler karşılığı):
// doğrulama → kullanıcı arama → şifre doğrulama → birincil üyelik → tenant →
// token üretimi. Kimlik bilgisi hatalarının tümü aynı "Invalid credentials."
// yanıtına eşlenir ki e-posta varlığı sızdırılmasın.
type LoginHandler struct {
	users       UserRepository
	memberships MembershipRepository
	tenants     TenantRepository
	tokens      TokenService
}

// NewLoginHandler, bağımlılıklarıyla yeni bir giriş handler'ı oluşturur.
func NewLoginHandler(users UserRepository, memberships MembershipRepository, tenants TenantRepository, tokens TokenService) *LoginHandler {
	return &LoginHandler{users: users, memberships: memberships, tenants: tenants, tokens: tokens}
}

// Execute, giriş komutunu işler ve başarılıysa LoginResult döner.
func (h *LoginHandler) Execute(ctx context.Context, cmd LoginCommand) sharedkernel.ResultOf[LoginResult] {
	if verr := ValidateLoginCommand(cmd); verr != nil {
		return sharedkernel.FailOf[LoginResult](verr)
	}

	user, err := h.users.GetByEmail(ctx, strings.ToLower(strings.TrimSpace(cmd.Email)))
	if err != nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewInternalError(err.Error()))
	}
	if user == nil || !users.VerifyPassword(cmd.Password, user.PasswordHash) {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewUnauthorizedError("Invalid credentials."))
	}

	membership, err := h.memberships.GetPrimaryForUser(ctx, user.ID)
	if err != nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewInternalError(err.Error()))
	}
	if membership == nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewUnauthorizedError("User is not assigned to a tenant."))
	}

	tenant, err := h.tenants.GetByID(ctx, membership.TenantID)
	if err != nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewInternalError(err.Error()))
	}
	if tenant == nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewUnauthorizedError("Tenant not found."))
	}

	token, expiresAt, err := h.tokens.Generate(user, tenant)
	if err != nil {
		return sharedkernel.FailOf[LoginResult](sharedkernel.NewInternalError(err.Error()))
	}
	return sharedkernel.OkOf(LoginResult{
		Token:     token,
		ExpiresAt: expiresAt,
		User:      UserDto{ID: user.ID, Email: user.Email, Name: user.Name},
		Tenant:    TenantDto{ID: tenant.ID, Name: tenant.Name},
	})
}
